src/uuid/uuid.nim

```diff
@@ -0,0 +1,67 @@
+import std/[strformat]
+import pkg/[uuids]
+import ".."/[logger]
+
+proc uuid*(n: Positive) =
+  ## Writes `n` version 4 UUIDs to stdout. Outputs only 1000 UUIDs if `n` is
+  ## greater than 1000.
+  const maxNumUuids = 1000
+  if n > maxNumUuids:
+    logNormal &"The UUID output limit is 1000, but {n} UUIDs were requested."
+    logNormal &"Outputting {maxNumUuids} UUIDs:"
+  for i in 1 .. min(n, maxNumUuids):
+    stdout.writeLine genUUID()
+
+func isValidUuidV4*(s: string): bool =
+  ## Returns `true` if `s` is a valid version 4 UUID (compliant with RFC 4122)
+  ## in the canonical textual representation.
+  ##
+  ## This func is equivalent to using `re.match` with the below regex pattern:
+  ##
+  ## `^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
+  ##
+  ## However, this func's implementation is faster and doesn't add a PCRE
+  ## dependency.
+  const
+    Hex = {'0'..'9', 'a'..'f'}
+    Separator = '-'
+    Version = '4'
+    Variant = {'8', '9', 'a', 'b'}
+  result =
+    s.len == 36 and
+    s[0] in Hex and
+    s[1] in Hex and
+    s[2] in Hex and
+    s[3] in Hex and
+    s[4] in Hex and
+    s[5] in Hex and
+    s[6] in Hex and
+    s[7] in Hex and
```

Comment:
Did you not use slices here for performance? Lot of repetitive code. I don't necessarily mind that here though, just curious.

Reply:
> Did you not use slices here for performance?

indeed. This is faster than something involving `s[0..7]`, because slicing in Nim involves a copy. It unfortunately isn't an alias for `toOpenArray(s, 0, 7)`.

`toOpenArray` was introduced in Nim 0.19.0 (September 2018), but making no-copy more ergonomic is still being worked on.

I also prefer the PR implementation to something like
```Nim
const ValidUuidV4 = "01234567-9012-4567-9012-456789012345"
doAssert ValidUuidV4.len == 36

func uuidIndices: (seq[int], seq[int]) =
  for i, c in ValidUuidV4:
    if c == '-':
      result[0].add i
    else:
      result[1].add i

const (HyphenIndices, HexIndices) = uuidIndices()

func isValidUuidV4*(s: string): bool =
  const
    Hex = {'0'..'9', 'a'..'f'}
    Separator = '-'
    Version = '4'
    Variant = {'8', '9', 'a', 'b'}

  if s.len == 36 and s[14] == Version and s[19] in Variant:
    for i in HexIndices:
      if s[i] notin Hex:
        return false
    for i in HyphenIndices:
      if s[i] != Separator:
        return false
    result = true
  else:
    return false
```

It's easy to write a macro and reduce the duplication. But I'd just make it generate the `func` in this PR, and the code would be less obvious.

Reply:
Yeah, let's just keep it this way